### PR TITLE
switch to meson build; stop bundling dependencies

### PR DESCRIPTION
**Pull request by mikolajpp**
This is the first working version of meson build system as discussed in #932 
Some small changes has to be done to accommodate non-linux systems.

### CODE REVIEW

**Comment by mikolajpp** [comment]
Since libuv is unbundled and libuv version in travis is outdated, travis scripts needs to build 
recent libuv from source.

**Comment by mikolajpp** [comment]
I have added support for BSD and OSX, though osx build should be tested by someone with access to mac.

**Comment by mikolajpp** [comment]
This adds debug/fast build via configuration options and updates README with short guide to using meson and ninja

**Comment by mikolajpp** [comment]
This provides working and tested (at least on @Fang- mac) meson build for osX.

**Comment by mikolajpp** [comment]
Looking at it, this branch has become a bit messy. If preferred, I am open to making a several commit in size branch against a master for cleanliness...

**Comment by mikolajpp** [comment]
And this moves dependencies onto upstream libscrypt and ed25519 versions.

**Comment by mikolajpp** [comment]
And this moves onto cleaned up repo of murmur3

**Comment by mikolajpp** [comment]
Okay, barred some bugs or corner cases, I consider this to be testable. 
Please try it out. 

This should be possible to be bundled into various distros. Hopefully we will soon get urbit into some major repos, such as ubuntu, archlinux, fedora and debian

File: meson.build [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+else
+  error('Little endian encoding required')
+endif
+
+configure_file(input : 'include/config.h.in',
+               output : 'config.h',
+               configuration : conf_data)
+
+# We expect these libs to supplied with the distribution
+openssl_dep = dependency('openssl')
```
Comment (belisarius222) [inline meson.build]: Do we not need to pin the versions of openssl, libuv, and curl? It feels a little dangerous to me to let them float based on the platform.

**Comment by belisarius222** [comment]
On my ubuntu 16.04 laptop, the meson build command errored out when I followed the instructions in the new README:

```
ted@novgorod:~$ git clone git@github.com:mikolajpp/urbit.git mikurbit
ted@novgorod:~$ cd mikurbit/
ted@novgorod:~/mikurbit$ sudo apt install meson
ted@novgorod:~/mikurbit$ git submodule init
ted@novgorod:~/mikurbit$ git submodule update
ted@novgorod:~/mikurbit$ meson build
Error during basic setup:

[Errno 2] No such file or directory: '/home/ted/mikurbit/build'
ted@novgorod:~/mikurbit$ mkdir build
ted@novgorod:~/mikurbit$ meson build
The Meson build system
Version: 0.29.0
Source dir: /home/ted/mikurbit
Build dir: /home/ted/mikurbit/build
Build type: native build
Build machine cpu family: x86_64
Build machine cpu: x86_64
Project name: urbit
Native c compiler: cc (gcc 5.4.0-6ubuntu1)

Meson encountered an error in file meson.build, line 224, column 2:
Unknown method "find_library" in object.
ted@novgorod:~/mikurbit$ 
```

**Comment by mikolajpp** [comment]
I am afraid your meson version is too old. I am going to put a constraint then in the meson.build.

About openssl versions, yes I guess we should pin them (as well as pin other versions).

I am unsure though, which one should be chosen. Also I don't think the old makefile pinned anything. It just took what was in the   system. I think for the stuff we still bundle, we can set version to what we bundle currently.

For system stuff I am going to check latest Ubuntu LTS and use them as reference. But Ubuntu 16 is definitely too old to support, if not just for the meson version. You need at least 0.40.

**Comment by mikolajpp** [comment]
OK, actually Ubuntu 16 Lts is OK.
You just need to install updated meson from xenial-backports.

**Comment by mikolajpp** [comment]
Okay, now we pin all dependencies. 
If you update meson, we support Ubuntu LTS 16.

**Comment by belisarius222** [comment]
It's weird the current Ubuntu LTS isn't directly supported by meson ... @mikolajpp what are the steps to install meson from xenial-backports?

**Comment by mikolajpp** [comment]
@belisarius222 I can't give you verified instructions, since I don't use Ubuntu, but I believe you have to enable xenial-backports ppa.

I don't know how ubuntu decides versions, but case in point it probably needed some newer libraries, like modern python...

**Comment by mikolajpp** [comment]
This: https://help.ubuntu.com/community/UbuntuBackports should set you

**Comment by mikolajpp** [comment]
Any update on this? 

Unsure if I can just move on to other stuff or expect other scrutiny ;-)

I do plan to develop against meson build branch though...

**Comment by jfranklin9000** [comment]
Now that `outside/anachronism` is gone `ragel` is no longer needed as a dependency.

In fact, the file produced by `ragel` (`outside/anachronism/src/parser.c`) has been checked into the repository so `ragel` hasn't even been used in a while.

```
git log --follow outside/anachronism/src/parser.c
commit 35fc5494114eff9801584461559ac45ffe0c8875
Author: C. Guy Yarvin <cgyarvin@gmail.com>
Date:   Wed Jun 3 16:35:54 2015 -0700

    Remove warnings.
```

**Comment by mikolajpp** [comment]
We don't require ragel in the meson build?

I removed any mentions of ragel from readme and travis build then.

**Comment by jfranklin9000** [comment]
On CentOS7 I had `libuv` installed.  I got this error message during `meson build`:
```
Meson encountered an error in file meson.build, line 262, column 0:
Native dependency 'libuv' not found
```
`yum install libuv-devel` solved this.  Maybe a note about devel versions of packages should go in `README.md`.

`yum install meson` on CentOS7 installs `0.42.1`. 
```
$ meson -v
0.42.1
```
But:
```
$ ninja
bash: ninja: command not found
```
Huh?

**Comment by mikolajpp** [comment]
Yes, ninja is also required. I thought it should be pulled with meson, but if not, I guess you need to install it separately. Maybe this also needs to go into README

**Comment by jfranklin9000** [comment]
CentOS7 doesn't have a `ninja` package but it can be built from source (https://ninja-build.org/):
```
git clone git://github.com/ninja-build/ninja.git && cd ninja
git checkout release
./configure.py --bootstrap
sudo cp ninja /usr/bin
```
I was able to complete the `meson` build of `urbit` with this.

**Comment by mikolajpp** [comment]
Great. I have managed to fix travis build. The outdated ubuntu really is a pain...

**Comment by mikolajpp** [comment]
Just so we do not forget. The dependencies which reside in my github, and which enable meson builds for stuff usually not found in distros needs to be moved under @urbit governance...

**Comment by belisarius222** [comment]
@mikolajpp Could I give you a call? I'd like to learn more about how you're thinking about this. Feel free to email me at ted@tlon.io so we can coordinate.

**Comment by mikolajpp** [comment]
@belisarius222 Sure thing.

**Comment by paulp** [comment]
I can confirm the OSX build works, to a first approximation anyway. I built the urbit binary and ran a comet without anything unexpected happening.

**Comment by joemfb** [comment]
This branch works well for me too (on El Capitan, after installing `python3` and `ninja` with homebrew).

Once this has been tested on relevant platforms and merged, will the Makefile be going away? Or will both build systems be maintained for the time being? (I'm asking specifically with regards to #942)

**Comment by mikolajpp** [comment]
In this branch, I have already removed Makefile and old build system. It does not make much sense to keep it, as the dependency management scheme is different, and we would have duplicate source code. 

But of course this is intrusion of authority ;-)
Someone with power has to weigh on this.

**Comment by joemfb** [comment]
Thanks! Not sure how I missed 43cd572 ...

**Comment by belisarius222** [comment]
I’ve discussed this with @eglaysher and @cgyarvin. We need Urbit to use an industry-standard build system, and while meson may become that someday, it still hasn’t reached version 1.0 and isn’t fully supported on the current Ubuntu LTS, which is a large percentage of Urbit’s install base.

You’ve done good work here! Cleaning up the makefile is a major step in the right direction, and if we ever want to make it into apt-get et. al., we have to stop bundling packages, so these are important contributions.

@mikolajpp How hard do you think it would be to switch this to CMake (ideally with a ninja backend, according to @eglaysher)?

**Comment by eglaysher** [comment]
I spent the morning exploring how hard it would be to switch to cmake. It's not exactly trivial. Meson has some nice features regarding the fallback versions of libraries.

I then took a close look at the meson build. I think we could actually use meson with some minor modifications to the build. Xenial, the last LTS, and one of our large customer targets, provides meson 0.29. Your meson file will work if you:

- Modify the version at the top to 0.29.
- Change every instance of `meson.get_compiler('c').find_library(` to `find_library(`
- Change the directions to include that you have to hand make the build/ dir with that version of meson.

I think with those modifications, this whole approach is acceptable.

**Comment by eglaysher** [comment]
I put up changes that make this work with meson on xenial here: https://github.com/eglaysher/urbit/tree/old-meson-version

**Comment by paulp** [comment]
@belisarius222 For it to run out of the box on the current Ubuntu LTS strikes me as a clearly non-negotiable requirement, but "industry-standard build system" is less obviously necessary (and less obviously defined.) I admit I only heard of meson for the first time because of this PR, but the improvements here are material, immediate, and apparently not easily transferred to another build system.

So FWIW, now that @eglaysher has dealt with ubuntu (and this from someone who shies away from non-standard tools wherever possible) I hope this PR isn't sidelined over a perceived failure to utilize standard tooling.

**Comment by belisarius222** [comment]
@paulp I agree that this is a clear improvement in several dimensions. @eglaysher, @ohAitch, and I discussed this in more detail today, and after @eglaysher explored what it would be like to try to convert this work to CMake, it looks like we'll be merging this using meson. It has support for some features that are quite difficult and cumbersome in CMake (@eglaysher knows more specifics than I do), to such an extent that it looks like we'll be better off using meson overall, even though it's still somewhat experimental.

**Comment by cgyarvin** [comment]
Agreed.  Great work by everyone on this.

**Comment by mikolajpp** [comment]
As has been said, indeed CMAKE is more cumbersome than meson, which is part of the reason I went straight for meson, having some unpleasant experiences with cmake even for simple projects. 

If we can make it to work with Ubuntu LTS, all the better. 

In fact, we could even make it so that config file detects meson version, and uses newer meson routines if available, and fallbacks to the old one if no new meson version is available.

If this is preferable to just downgrading meson config regardless of version, I can adapt our current config to this scheme over the weekend.

**Comment by cgyarvin** [comment]
That sounds ideal, Mikolaj.  That kind of historical cruft is the least bad historical cruft — always easy to know how and when to delete it.

**Comment by eglaysher** [comment]
I was going to start merging this code as is plus my changes for Xenial, but it looks like newer versions of meson just error on previous versions.

That's not a good sign for meson's stability,  but if you can get this working out of the box with different versions by checking meson's version at run-time, I'll still merge this.

**Comment by mikolajpp** [comment]
Anyway, we can detect that by adding a few if's. Please check if this works for you. I checked it on my box with meson 0.29.0, and meson 0.45.0. 

We support meson legacy up to 0.29.0, otherwise we ask user to update to at least 0.40.0. 
We sure don't want to handle all versions in between. Users running on Ubuntu LTS are covered with support for meson 0.29. All other users are expected to be able either to download this version of meson (which is easy through pip), or run on something newer able to sustain version of meson at least 0.40.0.

**Comment by mikolajpp** [comment]
I have also added instructions on running meson legacy to README.
Everyone running on Ubuntu LTS is welcome to try them out, to see if it indeed works as advertised.

**Comment by paulp** [comment]
Here's a Dockerfile toward showing all is kosher on ubuntu LTS. I'm not using the legacy meson because it's easy to upgrade with pip, though I'm pretty sure that works fine as well. I'm running a moon through this, looks good so far.

This docker build is on docker hub and can be run by anyone, here's a sample invocation which exposes ports to localhost and mounts a volume at `/urbit` so everything is persistent.

```bash
docker run \
  --volume /local/inst/urbit:/urbit \
  -p 127.0.0.1:8080:8080 \
  -p 127.0.0.1:8443:8443 \
  -ti paulp/urbit:latest \
  <your arguments to urbit here>
```
Here's the Dockerfile. This is very "of the moment" as it literally checks out and builds this pull request - it's not tied to a branch.
```Dockerfile
FROM ubuntu:latest

ENV DEBIAN_FRONTEND="noninteractive"
ENV BUILD_REF="pull/934/head"
ENV LIBS="libcurl4-gnutls-dev libgmp3-dev libncurses5-dev libsigsegv-dev libssl-dev libuv1-dev"
ENV PACKAGES="g++ git meson pkg-config python3-pip re2c"
ENV APTGET="apt-get -o=Dpkg::Use-Pty=0 -y --no-install-recommends"

ENV MESON_OPTIONS="-Ddebug=true" \
    LANG="C.UTF-8" \
    LC_ALL="C.UTF-8"

WORKDIR /root
RUN      $APTGET -q update \
      && $APTGET -q install $LIBS $PACKAGES \
      && rm -rf /var/lib/apt/lists/*
RUN      pip3 install --upgrade pip setuptools \
      && pip3 install -I meson \
      && echo meson version $(meson --version) \
      && git clone https://github.com/urbit/urbit --depth=1 --no-checkout

WORKDIR ./urbit
RUN      git --version \
      && git fetch origin $BUILD_REF:meson \
      && git checkout meson \
      && git submodule init \
      && git submodule update --depth=1 \
      && meson ../build

WORKDIR ../build
RUN    meson configure $MESON_OPTIONS \
    && ninja \
    && rm -rf /root/urbit

WORKDIR /urbit
ENTRYPOINT [ "/root/build/urbit" ]
```

**Comment by cgyarvin** [comment]
Neat!  Unfortunately I had given my heart to Urbit before the rise of the
great meta-Unix that is Docker and friends, but I like meta-Unix and on
balance it's a good thing.  It might be your father's OS but at least it's
not your grandfather's.

One thing to keep in mind, however, when you use Urbit on a container
system: Urbit is of course a single-level store, which makes it a storage
system, which is contrary to the intentions of the modern meta-Unix, which
really wants its containers to be transient and disposable (the 12-step
app, or 12-factor, or whatever) and access storage as a service.  It is
possible in theory to make Urbit access storage as a service (ideally it
should have a Kafka backend, or whatever) but this is not fully
implemented.

So you are pushing hard on the container filesystem as regards both
reliability and latency.  Due to my great age and silvered back, I am not
enough of a container expert to know how to approach this problem -- I just
know it is a problem, or at least might be.  I always get very nervous when
using a tool for something that isn't the same thing most people use it for.

Also: if you are serious about this dockerfile, I'd probably have it pull
arvo as well, so you can (a) run a fake zod and (b) browse around the
source tree trivially.  (I often go back and forth on whether urbit/urbit,
which is just the vere interpreter, and urbit/arvo, where all the real
semantic action is, should be renamed/unified/whatever.  Probably the
latter.)

**Comment by paulp** [comment]
Docker's advantages include isolation, testability, explication of dependencies, lowering the barrier to entry for the casually interested to see it in practice. These advantages would remain even were there something about docker which made it especially unsuitable for a long-running urbit.

It isn't my experience that docker containers which mount volumes from the host have filesystem reliability issues, and there is so much virtualization in modern environments it's not clear to me why we should single out this layer as especially risky.

I'm serious about the Dockerfile in the sense of seriously suggesting there be continuous testing and that docker is a valuable tool for providing end-to-end assurance about what is working. I'm not yet volunteering to make that a reality, but the Dockerfile above would be a great start for anyone who did.

**Comment by cgyarvin** [comment]
All very true.  However: I wouldn't say "filesystem reliability issues" in
the sense of just, like, *randomly losing data*, but a quick search turns
up stuff like this:

https://about.zoosk.com/en/engineering-blog/test-databases-docker-containers/

TLDR, fsync() on Docker is known to -- or was known to two years ago --
have execrable performance.  This means it is frequently turned off, etc.
You are getting off the beaten trail rapidly.  Moreover, as this post
notes, syncing a file on Unix is not as simple as it seems:

http://blog.httrack.com/blog/2013/11/15/everything-you-always-wanted-to-know-about-fsync/

Awful shit will happen to your urbit if it forgets that it didn't send an
ack, etc.  Every packet is a transaction.  Your inbound packet has to be
persisted before the response to it can be sent.  Relatively few systems
operate under this kind of constraint.

And in fact, the present Urbit event system really isn't 100% ACID -- I
don't trust its syncs at all.  ACID is hard.  That's why we have a new
event system waiting in the wings, that mostly just needs to be tested into
existence (I don't believe in it either, but I could be made to believe in
it).  So, worst-case scenario, we get a bunch of people running Urbit on
Docker and then have to be like: actually, there's no reliable and
efficient way to run Urbit on Docker, because fsync().

None of these things are unsolvable problems.  I think.  It's just like: we
have a lot of problems at the C layer and not a lot of people to throw at
it at present.  So "let's expand our deployment model to also run on
Docker" never reached the top of the stack.  Now that much of the work has
already been done *and* the result is this simple and clean, however... :-)

**Comment by paulp** [comment]
I am not without sympathy about problems introduced by containers. And syncing a file on unix has never seemed simple.

**Comment by paulp** [comment]
Here's an alternate Dockerfile which uses alpine linux instead of ubuntu. The image size drops from 416MB to 52.8MB. It's also on docker hub as `paulp/urbit-alpine`.

```Dockerfile
FROM alpine:latest

ENV TERM="xterm-256color" \
    APK="apk add --no-cache" \
    BUILD_REF="pull/934/head" \
    MESON_OPTIONS=""

WORKDIR /root
RUN      echo "@edge http://nl.alpinelinux.org/alpine/edge/testing" >> /etc/apk/repositories \
      && $APK --upgrade apk-tools@edge \
      && $APK curl-dev gmp-dev libuv-dev ncurses-dev openssl-dev libsigsegv-dev@edge \
      && $APK --virtual .build-deps git g++ meson re2c \
      && git clone https://github.com/urbit/urbit --depth=1 --no-checkout \
      && cd urbit \
      && git fetch origin $BUILD_REF:meson \
      && git checkout meson \
      && git submodule init \
      && git submodule update --depth=1 \
      && meson ../build \
      && cd ../build \
      && meson configure $MESON_OPTIONS \
      && ninja \
      && rm -rf /root/urbit \
      && apk del .build-deps

WORKDIR /urbit
ENTRYPOINT [ "/root/build/urbit" ]
```

File: .travis.yml [hunk 2]
```diff
@@ -1,38 +1,52 @@
...
-# notifications: 
-#   email: false 
+# notifications:
+#   email: false
+
+before_install:
+  - wget https://github.com/ninja-build/ninja/releases/download/v1.8.2/ninja-linux.zip
+  - unzip ninja-linux.zip
+  - sudo mv ninja /usr/bin/
+  - wget https://github.com/libuv/libuv/archive/v1.19.2.tar.gz
```
Comment (ohAitch) [inline .travis.yml]: Is the correct version of `libuv` neither a submodule nor available as a package on LTS? I'm leery of installing things informally like this, for the CI it's probably fine but I expect it translates to everyone trying to build urbit also having to resolve these dependencies.
